packages/adapter/src/lib/adapter/userInterfaceMessagingController.ts

```diff
@@ -0,0 +1,184 @@
+import type { Client as StatesInRedisClient } from '@iobroker/db-states-redis';
+import type { AdapterClass } from './adapter';
+import type {
+    UserInterfaceClientSubscribeHandler,
+    UserInterfaceClientSubscribeReturnType,
+    UserInterfaceClientUnsubscribeHandler
+} from '../_Types';
+import { clearTimeout } from 'timers';
+
+export interface HeartbeatTimer {
+    /** The actual timer */
+    timer: NodeJS.Timeout;
+    /** The heartbeat interval */
+    heartbeat: number;
+}
+
+export interface MessagingControllerOptions {
+    /** The adapter using this messaging controller */
+    adapter: AdapterClass;
+    /** Callback to call if successfully subscribed */
+    subscribeCallback?: UserInterfaceClientSubscribeHandler;
+    /** Callback to call if successfully unsubscribed */
+    unsubscribeCallback?: UserInterfaceClientUnsubscribeHandler;
+}
+
+export interface SendToClientOptions {
+    /** ID of the client to send the message to */
+    clientId: string;
+    /** Data to send to the client */
+    data: unknown;
+    /** The states db */
+    states: StatesInRedisClient;
+}
+
+export interface ClientHandler {
+    /** The session id of the client connection */
+    sid: string;
+    /** Name of the subscriber */
+    from: string;
+    /** Timestamp of subscription */
+    ts: number;
+    /** Individual type which can be specified */
+    type: string;
+}
+
+export class UserInterfaceMessagingController {
+    /** The adapter using this messaging controller */
+    private readonly adapter: AdapterClass;
+    /** Callback to call if successfully subscribed */
+    private readonly unsubscribeCallback?: UserInterfaceClientUnsubscribeHandler;
+    /** Callback to call if successfully unsubscribed */
+    private readonly subscribeCallback?: UserInterfaceClientSubscribeHandler;
+    /** All currently registered client handlers */
+    private readonly handlers = new Map<string, ClientHandler>();
+    /** Collection of current heartbeat timers */
+    private heartbeatTimers = new Map<string, HeartbeatTimer>();
+    constructor(options: MessagingControllerOptions) {
+        const { adapter, unsubscribeCallback, subscribeCallback } = options;
+
+        this.adapter = adapter;
+        this.unsubscribeCallback = unsubscribeCallback;
+        this.subscribeCallback = subscribeCallback;
+    }
+
+    /**
+     * Send a message to the client configured by the handler
+     *
+     * @param options Data and client information
+     */
+    sendToClient(options: SendToClientOptions): Promise<void> {
+        const { states, clientId, data } = options;
+
+        if (!this.handlers.has(clientId)) {
+            throw new Error(`Client "${clientId}" is not registered`);
+        }
+
+        const handler = this.handlers.get(clientId)!;
+
+        return states.pushMessage(handler.from, {
+            command: 'im',
+            message: { m: handler.type, s: handler.sid, d: data },
```

Comment:
Can we make the property names and the command more meaningful @GermanBluefox? Maybe I also misinterpreted, that commadn needs to be `im` like in the example at https://github.com/ioBroker/ioBroker.socket-classes#gui-subscribes

Reply:
No. I save here the traffic. Because I send with it the images to GUI and additional 100 bytes in every frame could be easy transformed into 1k by 10 frames per second.

Reply:
And exactly because of these short names I want to hide it into adapter.js, so the dev does not use these short names

Reply:
Okay, if you say it mattered in your tests we can keep it